### PR TITLE
vscode devcontainer

### PR DESCRIPTION
**Pull request by luwol03**
I already shared this in this in https://github.com/inventree/InvenTree/pull/3429#issuecomment-1205407752 . I used this setup for my latest two pr #3451 , #3473 and it works great.

@SchrodingersGat what do you think of that? What IDE are you using locally?

I will create this PR as draft and if the team likes this, I'll do some improvements I have in mind.

<a href="https://gitpod.io/#https://github.com/inventree/InvenTree/pull/3474"></a>

### CODE REVIEW

**Comment by SchrodingersGat** [comment]
@luwol03 I might need to get my head around this.

I use vscode for dev, with InvenTree running in docker-desktop:

This means that I can shell into the running containers via vscode, it's very clean! It also means that I am always running using the development docker environment.

**Comment by luwol03** [comment]
@SchrodingersGat I improved it even a step further, maybe you will also like the debugger and the available syntax highlighting. I recorded a little video to show the features. Note, that the first time start will take some longer time, as the requirements needs to be fetched and the docker file be build.

https://user-images.githubusercontent.com/60048565/183087281-7df51a40-4743-41d6-b1f5-fbf19ad37309.mp4

image

File: .vscode/launch.json
Comment (matmair) [inline .vscode/launch.json:16]: Suggestion: Add a second config with justMyCode disabled - we tend to call a lot of 3rd party packages, debugging is nearly useless without looking into those.

File: .pre-commit-config.yaml
Comment (matmair) [inline .pre-commit-config.yaml:19]: Please Change: This is a dependency update and might lead to problems further down the line - please submit this in a new PR so we can label and reproduce it easier.

File: .gitignore
Comment (matmair) [inline .gitignore:76]: Question: Won't this lead to a lot of changes in git? I have 50+ extensions in my full InvenTree dev enviroment

File: .devcontainer/devcontainer.json
Comment (matmair) [inline .devcontainer/devcontainer.json:48]: Suggestion: Nothing from the requirements should be needed to call inv update, if you turn around the order we code safe one action probably:
```suggestion
  "postCreateCommand": "python3 -m venv venv && . venv/bin/activate && cd /workspaces/InvenTree && pip install invoke && inv update && mkdir -p dev && inv setup-dev",
```

File: .devcontainer/devcontainer.json
Comment (matmair) [inline .devcontainer/devcontainer.json:40]: Suggestion: I use a coverage extension to quickly check if new tests cover all new code @luwol03 @SchrodingersGat would that be a reasonable addition? We would need to add `coverage.py` to the dev-reqs then probably.

**Review by matmair — CHANGES_REQUESTED** [review]
Looks good @luwol03 ; Please move the dependency update out the rest will be a great improvement.

**Comment by luwol03** [comment]
@matmair thank you for your review. Ill take a deeper look later.

Good idea, please suggest useful extensions (with their identifier) we should add.

File: .gitignore
Comment (luwol03) [reply .gitignore:76]: I just used the template linked above, but we can comment this and uncomment as we need it. I guess in `extensions.json` we can specify extensions for usage without the devcontainer, as they are already specified in the `devcontainer.json`.

File: .pre-commit-config.yaml
Comment (luwol03) [reply .pre-commit-config.yaml:19]: sorry, forgot to remove that file from committing. It was changed by some of the auto update invokes.

File: .devcontainer/devcontainer.json
Comment (luwol03) [reply .devcontainer/devcontainer.json:40]: Do you know how to improve the testing setup as its pretty bad marked in https://github.com/inventree/InvenTree/issues/3384#issuecomment-1206799712 ? I tried to setup the test extension, but didn't got it working, so I sticked with the CLI and prints.

**Review by matmair — APPROVED** [review]

**Comment by luwol03** [comment]
@matmair can you share the identifiers of the extensions you recommend?

File: .devcontainer/devcontainer.json
Comment (matmair) [reply .devcontainer/devcontainer.json]: @luwol03 we are using the default solution for Django. It is the same as in most projects.
The test extension normally auto-discovers test if you chose the right framework at the start (unittest or pytest).

I really do not care how the code gets developed as long as the prints do not make into production code. My suggestion would be to use the debug console or start the Django shell and develop the tests in there. Most test consist of short one-liners so that usually works fine.

Comment (luwol03) [reply .devcontainer/devcontainer.json]: Which is the correct one? unites or pytest?

**Comment by matmair** [comment]
@luwol03 Github codespaces are actually pretty expensive compared to gitpod in my experience.

**Comment by matmair** [comment]
@luwol03 the changes to enable my testing setup in devcontainers turned out to be quite involved.
I think it is better to add some docs to CONTRIBUTING for this changeset and merge it before this PR gets out of hand. I will make a separate PR once I figure out how to streamline the whole thing (would currently need 5+ extensions, a lot of setup and an old node.js version)

File: .devcontainer/devcontainer.json
Comment (matmair) [reply .devcontainer/devcontainer.json]: @luwol03 they support the same call API. While we use unittest, both will work.

**Comment by luwol03** [comment]
> @luwol03 Github codespaces are actually pretty expensive compared to gitpod in my experience.

Yeah, just wanted to mention that one if somebody will need it.


> @luwol03 the changes to enable my testing setup in devcontainers turned out to be quite involved.
> I think it is better to add some docs to CONTRIBUTING for this changeset and merge it before this PR gets out of hand. I will make a separate PR once I figure out how to streamline the whole thing (would currently need 5+ extensions, a lot of setup and an old node.js version)

So this PR can be marked as ready to review for your side and you'll add your extensions in another PR?

File: .devcontainer/Dockerfile
Comment (SchrodingersGat) [inline .devcontainer/Dockerfile:18]: For a fully compatible setup you should install all required system packages 

https://github.com/inventree/InvenTree/blob/e1d3392f3743f7813c0aba0705223f1442526ca7/Dockerfile#L61

File: .devcontainer/devcontainer.json
Comment (SchrodingersGat) [inline .devcontainer/devcontainer.json:63]: I'd prefer to keep all these directories under the same directory. Currently you have them split between `./dev/` and `./inventree-data/`.

For the development docker image we use `./data/media/` `./data/static/` (etc)

**Comment by SchrodingersGat** [comment]
@luwol03 while I am hesitant to have "too many" ways to achieve the same thing (in this case a development setup) this does seem like a great way to get up and running very quickly.

There are some downsides I can see:

- Limited to sqlite (fine for simple dev)
- Doesn't run the background worker process (again, fine for simple dev)

I'd be happy to see this as a "get a quick dev environment setup" tutorial (documented of course), with a lead-in to the full development setup e.g. "for a more fully featured development setup, refer to the development docker guide" or something like that.

@matmair what do you think about this? How do you see this vscode integration fitting in with our current setup guide (we have 4 currently!)

File: .devcontainer/devcontainer.json
Comment (luwol03) [reply .devcontainer/devcontainer.json:63]: I copied this from the gitpod file. But you're right.

**Comment by luwol03** [comment]
@SchrodingersGat Thank you for your feedback.

You're not limited to sqlite. You can also configure a docker compose stack in the devcontainer.json and attach to a specified service. This would be helpful to setup a database, redis, ...

### Questions

1. How do I start the work?
2. How do we make a select which of the featured levels you'd like to start? Because I guess for first time contributor who has no idea about the setup, the current state is just fine, but for you as core developer, you need more different things like different database, worker, redis, ...

**Comment by SchrodingersGat** [comment]
I'm not sure how *useful* it is to fully flesh out with multiple containers etc. Even having the background worker *not* running is a good thing for development as it forces the processes to run in the server thread - great for debugging.

As for "which features should you include"... I don't know. Every new setup is another thing we need to document and support. 

I'm happy to promote this as a "developer setup" guide, and anything beyond is left to the user. 

The current setup guides do not really help with new developers who want to get stuck into coding. This would be a great addition in this regard. We should add a new setup guide to the docs specifically for developers. Your video would work well there I think

**Comment by luwol03** [comment]
If we want to use that video, I would like to make a new one which shows more available features. Maybe we should also wait until @matmair added some more extensions?

**Comment by luwol03** [comment]
@SchrodingersGat Please let me know what the next steps will be.

**Comment by SchrodingersGat** [comment]
@luwol03 I am currently running through the process of setting up on my system to replace my dev setup.

One thing I see is that we already *have* a dockerfile which you are essentially copying (the first part of at least).

Is it possible to inherit from the Dockerfile at the top level directory? That way if we ever have to update our docker file, at least we won't have to change two files?

**Comment by SchrodingersGat** [comment]
I ran into this error during the initial setup process:

image

Have you seen this?

File: .devcontainer/Dockerfile
Comment (SchrodingersGat) [inline .devcontainer/Dockerfile:43]: Does `/workspaces/InvenTree` have to be hardcoded here? Will this still work if the user calls the workspace something else? Or *must* the user load the workspace as `InvenTree`?

**Comment by SchrodingersGat** [comment]
@luwol03 I made the following patch with some small additions for a "cleaner" initial setup.

Please apply the patch and push the changes.

[devcontainer.diff.txt](https://github.com/inventree/InvenTree/files/9276193/devcontainer.diff.txt)

File: .devcontainer/Dockerfile
Comment (luwol03) [reply .devcontainer/Dockerfile:43]: This is the path in the container. If you want to change that you also have to specify something in the devcontainer.json which is not available in codespaces, but this should be the default.

**Comment by luwol03** [comment]
> Is it possible to inherit from the Dockerfile at the top level directory? That way if we ever have to update our docker file, at least we won't have to change two files?

I don't know of any way, because we need the vscode image as base. I'd be happy if you know any ways.


> @luwol03 I made the following patch with some small additions for a "cleaner" initial setup.
> Please apply the patch and push the changes.

Thank you for the additions. They look good. I'll apply it for you, but you can also push yourself as a maintainer to my branch. You need to checkout the refs/pulls/3348 ref I guess, but this is simpler with the gh cli, it's just gh pr checkout 3348 and then you're able to push to my branch.

**Comment by luwol03** [comment]
@SchrodingersGat There is one thing I have to notice here,
```diff
   // Use 'postCreateCommand' to run commands after the container is created.
-  "postCreateCommand": "mkdir -p /workspaces/InvenTree/dev/commandhistory && python3 -m venv venv && . venv/bin/activate && cd /workspaces/InvenTree && pip install invoke && inv update && mkdir -p dev && inv setup-dev",
+  "postCreateCommand": "mkdir -p /workspaces/InvenTree/dev/commandhistory && cd /workspaces/InvenTree && python3 -m venv dev/venv && . dev/venv/bin/activate && pip install invoke && inv update && inv setup-test --ignore-update --path dev/demo-dataset",
```

If we do it like this, we should make sure that `setup-test` don't run if already data exists in the database, because this command runs every time you remove you recreate your devcontainer (e.g. rebuild, delete all exited containers and then start the devcontainer again, ...). But I prefer it as a task (Use cmd pallet type `> Run task` press enter, select `setup-test` press ok, wait until complete)

**Comment by luwol03** [comment]
We can share this video if you want.

https://user-images.githubusercontent.com/60048565/183285495-77d9a32c-300f-4e23-b764-1678cfeb5e51.mp4

**Comment by SchrodingersGat** [comment]
> If we do it like this, we should make sure that setup-test don't run if already data exists in the database, because this command runs every time you remove you recreate your devcontainer (e.g. rebuild, delete all exited containers and then start the devcontainer again,

Ah good point. I guess it makes more sense to run the "demo data" setup as a separate action. 

The `setup-dev` command failed locally for me, I'll have to look into that some more.

**Comment by SchrodingersGat** [comment]
@luwol03 I'll have another look at this hopefully tomorrow :)

**Comment by luwol03** [comment]
@SchrodingersGat Can you please share the error you had with setup-dev?

**Comment by matmair** [comment]
Looks good. Don't wait for the extra extensions @luwol03 , I have to rewrite several tasks for it to work seamlessly.

**Comment by luwol03** [comment]
@matmair no, I'm not waiting. I already posted the video. Currently I'm waiting for @SchrodingersGat to answer my notices from his patch in https://github.com/inventree/InvenTree/pull/3474#issuecomment-1207351858 . If that's all resolved, I think we can merge this from my side.

**Comment by SchrodingersGat** [comment]
> Can you please share the error you had with setup-dev?

Looks like a permission error. Still see it when running the task.

image

File: .devcontainer/postCreateCommand.sh
Comment (SchrodingersGat) [inline .devcontainer/postCreateCommand.sh:14]: You're right that this "setup-test" should not be performed "every time" the container is built. We should suggest it as a one-time step after setup.

Can you revert this to the "setup-dev" command?

**Comment by SchrodingersGat** [comment]
Getting pretty close @luwol03 

Regarding the demo video, I would suggest a simplification to the workflow. There should be only three steps to get "up and running":

### Download / Build

Download (clone) code and build the devcontainer

### Demo Dataset

Install demo dataset with the `setup-test` task

### Debug Server

Launch the server in debug mode

**Comment by luwol03** [comment]
And what is with showing the needed extension?

**Comment by SchrodingersGat** [comment]
> And what is with showing the needed extension?

Sorry I don't understand what you are asking

**Comment by luwol03** [comment]
@SchrodingersGat The remote containers extension is needed for all that to work. I showed that at the beginning after cloning the repo in my last video.

**Comment by SchrodingersGat** [comment]
Ah understood. Yes that absolutely needs to be in there. I really liked the structure of your video. There are a number of smaller things that I guess I did not specify in my "simple three step plan" above :)

**Comment by luwol03** [comment]
Can you describe exactly what you want to be in the video? And another idea would be to add litte texts to the video like Requirements, building, debugging, ... (the chapters)

**Comment by SchrodingersGat** [comment]
@luwol03 I have opened a new issue on the docs site - https://github.com/inventree/inventree-docs/issues/350 - where we should discuss the documentation aspect on this. Let's keep this PR for the actual feature :)

**Comment by SchrodingersGat** [comment]
I think the only outstanding item here is to work out the `setup-dev` bug which was a permission issue. @luwol03  do you have any idea why it might have failed on my end?

I'm running vscode under windows 10 if that makes any difference.

**Comment by luwol03** [comment]
@SchrodingersGat Regarding to your bug, have you checked the file permissions in the container? The vscode user needs at least execute and read. Maybe find out which exact command that is and try running it directly instead through the post task.

I can only test on macos, and cannot reproduce that. I'll try that one more time later.

**Comment by SchrodingersGat** [comment]
@luwol03 I'll have a look into that further. I'm happy to accept this as is, might just be my setup.

Can you have a look at https://github.com/inventree/inventree-docs/issues/350 when you have some time?

**Comment by luwol03** [comment]
Oh nice, that this is merged. Yes, I already had a look at the other issue on the docs site. Great team btw.

**Pull request merged**
